### PR TITLE
chore: migrate to `sturdy_http`

### PR DESCRIPTION
**Pull request by btrautmann**
### 📰 Summary of changes

> What is the new functionality added in this PR?

This PR migrates all network interactions to `sturdy_http`. It's dependent on https://github.com/Betterment/sturdy_http/pull/3 merging first and being released.

### 🧪 Testing done

> What testing was added to cover the functionality added in this PR

Unit tests!

### CODE REVIEW

File: packages/test_track/test/networking/fakes/fake_http_client.dart [hunk 1]
```diff
@@ -1,13 +1,10 @@
...
 
-class FakeHttpClient extends HttpClient {
-  FakeHttpClient(
+class FakeSturdyHttp extends SturdyHttp {
+  FakeSturdyHttp(
     Charlatan charlatan,
   ) : super(
-          baseUrl: 'http://localhost',
-          customAdapter: charlatan.toFakeHttpClientAdapter(),
-          logger: FakeTestTrackLogger.withoutNetworkLogging(),
```
Comment (samandmoore) [inline packages/test_track/test/networking/fakes/fake_http_client.dart:11]: we don't need this logger thing anymore?

File: packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart [hunk 2]
```diff
@@ -1,31 +1,42 @@
...
-  Future onError(DioError err, ErrorInterceptorHandler handler) async {
+  Future<void> onError(
+      DioException err, ErrorInterceptorHandler handler) async {
     var extra =
         RetryOptions.fromRequestOptions(err.requestOptions) ?? _retryOptions;
     final isIdempotent =
-        err.requestOptions.extra[HttpClient.isIdempotentOptionsKey] as bool? ??
-            false;
+        err.requestOptions.extra[isIdempotentOptionsKey] as bool? ?? false;
+    final uppercased = StringBuffer()
```
Comment (samandmoore) [inline packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart:34]: we should probably call this pascalCased because it's not quite UPPERCASED it's more PascalCased, right?

Comment (samandmoore) [reply packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart:34]: or in simpler terms, we could called it `capitalized` as in `Capitalized`

Comment (btrautmann) [reply packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart:34]: Hehe yeah

File: packages/test_track/lib/src/domain/get_visitor_config.dart [hunk 3]
```diff
@@ -33,12 +34,20 @@ class GetVisitorConfig {
     required String visitorId,
     required AppVersionBuild appVersionBuild,
   }) async {
-    final response = await _client.get<Map<String, dynamic>>(
-      '/api/v4/apps/${appVersionBuild.appName}/versions/${appVersionBuild.version}/builds/${appVersionBuild.buildTimestamp}/visitors/$visitorId/config',
+    final appVisitorConfig = await _client.execute<Json, AppVisitorConfig>(
+      IdempotentGetRequest(
```
Comment (matt-kanovsky) [inline packages/test_track/lib/src/domain/get_visitor_config.dart:38]: i see that this adds the `isIdempotentOptionsKey: true` to the request--I presume this key is something consumed by this visitor config endpoint and does something on the backend?

File: packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart [hunk 4]
```diff
@@ -1,31 +1,42 @@
...
     var extra =
         RetryOptions.fromRequestOptions(err.requestOptions) ?? _retryOptions;
     final isIdempotent =
-        err.requestOptions.extra[HttpClient.isIdempotentOptionsKey] as bool? ??
-            false;
+        err.requestOptions.extra[isIdempotentOptionsKey] as bool? ?? false;
+    final originalMethod = err.requestOptions.method;
+    final capitalizedMethod = StringBuffer()
+      ..write(originalMethod.substring(0, 1).toUpperCase())
+      ..write(originalMethod.substring(1).toLowerCase());
```
Comment (matt-kanovsky) [inline packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart:37]: 😲

File: packages/test_track/lib/src/domain/get_visitor_config.dart (hunk 3 above)
Comment (btrautmann) [reply packages/test_track/lib/src/domain/get_visitor_config.dart:38]: That key is just used by the SDK to indicate requests that are idempotent (i.e can be invoked > 1 times without risk of producing incorrect state). It's used by the `RetryInterceptor` so it knows whether or not it can retry requests! I didn't bother documenting it primarily because it's not exported and only used internal to the SDK, but LMK if you think that's a bad idea 🙃

File: packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart (hunk 4 above)
Comment (btrautmann) [reply packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart:37]: We actually assign it to a `NetworkRequestType` which is just our `enum` representation of the HTTP method! We then use that type as the `type` parameter to `RawRequest`. Without `RawRequest` (introduced [here](https://github.com/Betterment/sturdy_http/pull/3)), we'd have to `switch` on the `NetworkRequestType` and produce either a `GetRequest`, `PostRequest`, `PutRequest` or `DeleteRequest`, all of which would contain the exact same metadata but with a different HTTP method!

File: packages/test_track/lib/src/domain/get_visitor_config.dart (hunk 3 above)
Comment (matt-kanovsky) [reply packages/test_track/lib/src/domain/get_visitor_config.dart:38]: cool! i think it's fine as is

File: packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart (hunk 4 above)
Comment (matt-kanovsky) [reply packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart:37]: oooooooh !!!! that's sick

File: packages/test_track/lib/src/domain/get_visitor_config.dart (hunk 3 above)
Comment (btrautmann) [reply packages/test_track/lib/src/domain/get_visitor_config.dart:38]: I documented it, since I think this convo is proof that it's not 100% apparent at the outset! So thanks for bringing it up 😄

File: packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart [hunk 5]
```diff
@@ -36,24 +47,59 @@ class RetryInterceptor extends Interceptor {
...
+                  requestOptions: err.requestOptions,
+                ),
+              ),
+              okNoContent: () => handler.resolve(
+                Response(
+                  statusCode: 204,
+                  requestOptions: err.requestOptions,
+                ),
+              ),
+              genericError: (_, __, error) => throw error ?? errorForRequest(),
```
Comment (matt-kanovsky) [inline packages/test_track/lib/src/networking/interceptors/retry_interceptor.dart:88]: in what scenario would we actually hit `errorForRequest`? can `error` ever be null? i know the `genericError` is the default case for when the error code falls through.

**Review by matt-kanovsky — APPROVED** [review]
domainlgtm :-)

**Review by matt-kanovsky — APPROVED** [review]
domainlgtm

**Review by matt-kanovsky — APPROVED** [review]
domainlgtm 😢 😢 😭 😭

**Review by matt-kanovsky — APPROVED** [review]
DOMAIN LGTM

File: pubspec.yaml
Comment (ClaireDavis) [inline pubspec.yaml:1]: is this top-level pubspec supposed to be here?

**Review by ClaireDavis — COMMENTED** [review]
one little question :)

File: pubspec.yaml
Comment (btrautmann) [reply pubspec.yaml:1]: Yeah that's for `melos`!

Comment (btrautmann) [reply pubspec.yaml:1]: https://melos.invertase.dev/getting-started#install-melos-in-the-workspace

Comment (ClaireDavis) [reply pubspec.yaml:1]: cool! :)

**Review by ClaireDavis — APPROVED** [review]
domain lgtm
